[reland] make intrusive_ptr as a pybind holder type

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#47586 [reland] make intrusive_ptr as a pybind holder type**

relanding PR of https://github.com/pytorch/pytorch/pull/44492, and add
additional Capsule related wrapping to ensure we still have the correct
type in pybind11 to resolve Capsule as torch._C.CapsuleType

Differential Revision: [D24822519](https://our.internmc.facebook.com/intern/diff/D24822519)

wanchaol committed:
[reland] make intrusive_ptr as a pybind holder type

relanding PR of https://github.com/pytorch/pytorch/pull/44492, and add
additional Capsule related wrapping to ensure we still have the correct
type in pybind11 to resolve Capsule as torch._C.CapsuleType

[ghstack-poisoned]

wanchaol committed:
Update on "[reland] make intrusive_ptr as a pybind holder type"


relanding PR of https://github.com/pytorch/pytorch/pull/44492, and add
additional Capsule related wrapping to ensure we still have the correct
type in pybind11 to resolve Capsule as torch._C.CapsuleType

Differential Revision: [D24822519](https://our.internmc.facebook.com/intern/diff/D24822519)

[ghstack-poisoned]

wanchaol committed:
Update on "[reland] make intrusive_ptr as a pybind holder type"


relanding PR of https://github.com/pytorch/pytorch/pull/44492, and add
additional Capsule related wrapping to ensure we still have the correct
type in pybind11 to resolve Capsule as torch._C.CapsuleType

Differential Revision: [D24822519](https://our.internmc.facebook.com/intern/diff/D24822519)

[ghstack-poisoned]

gmagogsfm reviewed (approved):
Nice! PR looks good to me. Those test failures seem irrelevant.

wanchaol committed:
Update on "[reland] make intrusive_ptr as a pybind holder type"


relanding PR of https://github.com/pytorch/pytorch/pull/44492, and add
additional Capsule related wrapping to ensure we still have the correct
type in pybind11 to resolve Capsule as torch._C.CapsuleType

Differential Revision: [D24822519](https://our.internmc.facebook.com/intern/diff/D24822519)

[ghstack-poisoned]

swolchok commented on c10/util/intrusive_ptr.h:
```diff
@@ -207,16 +225,37 @@ class intrusive_ptr final {
+      // and because we allow raising these values from 0, which retain_()
+      // has an assertion against.
+      ++target_->refcount_;
+      ++target_->weakcount_;
```
Doesn't this introduce the possibility of a memory leak? If `target` already has one intrusive_ptr referencing it, this will set weakcount_ to 2, meaning that ~intrusive_ptr will never be able to `delete target`.

wanchaol replied:
while it's possible, but this is a private constructor and it's not intended to use by users who are trying to create a intrusive_ptr from T*, the only way user can create intrusive_ptr is from `make_intrusive`, which is guaranteed that there's no refcount already referring to it.
